### DISCUSSION TITLE
Getting 500 error code when pushing (Channel "139da5a4-e3a7-4f54-9189-9db3ccc4d352" belongs to more than one group)

### DISCUSSION THREAD

**Discussion by anibal2j** [SagaHealthcareIT/mirthsync#71] [General]
Hi all,

I've been using mirthsync to pull the code from an existing Mirth Connect server and then pushing the code to github and it's working beautifully!

I'm now trying the next step, which is to push the configuration to another Mirth Connect server (to keep both in sync). I'm getting a 500 error when pushing and the error on the secondary server, i.e., the one I'm pushing too is showing:

Channel '139da5a4-e3a7-4f54-9189-9db3ccc4d352' belongs to more than one group

However, I can see that this channel doesn't belong to more than one group.  Is there anything I can do to troubleshoot this?

The full error is:

`
clojure.lang.ExceptionInfo: clj-http: status 500 {:cached nil, :request-time 229, :repeatable? false, :protocol-version {:name "HTTP", :major 1, :minor 1}, :streaming? true, :http-client #object[org.apache.http.impl.client.InternalHttpClient 0x44841b43 "org.apache.http.impl.client.InternalHttpClient@44841b43"], :chunked? false, :type :clj-http.client/unexceptional-status, :reason-phrase "Internal Server Error", :headers {"Access-Control-Allow-Headers" "Content-Type", "Content-Type" "application/xml", "Access-Control-Allow-Origin" "*", "X-Frame-Options" "DENY", "Strict-Transport-Security" "max-age=31536000; includeSubDomains", "Connection" "close", "Access-Control-Allow-Methods" "GET, POST, DELETE, PUT", "Date" "Mon, 20 May 2024 21:21:45 GMT", "Access-Control-Allow-Credentials" "false", "Content-Security-Policy" "frame-ancestors 'none'"}, :orig-content-encoding nil, :status 500, :length -1, :body "<com.mirth.connect.client.core.ControllerException>\n  <detailMessage>Channel &quot;139da5a4-e3a7-4f54-9189-9db3ccc4d352&quot; belongs to more than one group.</detailMessage>\n  <stackTrace>\n    <trace>com.mirth.connect.server.controllers.DefaultChannelController.updateChannelGroups(DefaultChannelController.java:837)</trace>\n    <trace>com.mirth.connect.server.api.servlets.ChannelGroupServlet.updateChannelGroups(ChannelGroupServlet.java:54)</trace>\n    <trace>sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)</trace>\n    <trace>sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)</trace>\n    <trace>sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)</trace>\n    <trace>java.lang.reflect.Method.invoke(Method.java:498)</trace>\n    <trace>com.mirth.connect.server.api.providers.MirthResourceInvocationHandlerProvider$1.invoke(MirthResourceInvocationHandlerProvider.java:219)</trace>\n    <trace>org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144)</trace>\n    <trace>org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161)</trace>\n    <trace>org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:205)</trace>\n    <trace>org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99)</trace>\n    <trace>org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389)</trace>\n    <trace>org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347)</trace>\n    <trace>org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102)</trace>\n    <trace>org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326)</trace>\n    <trace>org.glassfish.jersey.internal.Errors$1.call(Errors.java:271)</trace>\n    <trace>org.glassfish.jersey.internal.Errors$1.call(Errors.java:267)</trace>\n    <trace>org.glassfish.jersey.internal.Errors.process(Errors.java:315)</trace>\n    <trace>org.glassfish.jersey.internal.Errors.process(Errors.java:297)</trace>\n    <trace>org.glassfish.jersey.internal.Errors.process(Errors.java:267)</trace>\n    <trace>org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317)</trace>\n    <trace>org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305)</trace>\n    <trace>org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154)</trace>\n    <trace>org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:471)</trace>\n    <trace>org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:425)</trace>\n    <trace>org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:383)</trace>\n    <trace>org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:336)</trace>\n    <trace>org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:223)</trace>\n    <trace>org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)</trace>\n    <trace>org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1631)</trace>\n    <trace>com.mirth.connect.server.api.providers.StrictTransportSecurityFilter.doFilter(StrictTransportSecurityFilter.java:33)</trace>\n    <trace>org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)</trace>\n    <trace>org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)</trace>\n    <trace>com.mirth.connect.server.MethodFilter.doFilter(MethodFilter.java:37)</trace>\n    <trace>org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)</trace>\n    <trace>org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)</trace>\n    <trace>com.mirth.connect.server.api.providers.RequestedWithFilter.doFilter(RequestedWithFilter.java:53)</trace>\n    <trace>org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)</trace>\n    <trace>org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)</trace>\n    <trace>com.mirth.connect.server.api.providers.ClickjackingFilter.doFilter(ClickjackingFilter.java:44)</trace>\n    <trace>org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)</trace>\n    <trace>org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)</trace>\n    <trace>com.mirth.connect.server.api.providers.ApiOriginFilter.doFilter(ApiOriginFilter.java:71)</trace>\n    <trace>org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)</trace>\n    <trace>org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1601)</trace>\n    <trace>org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:548)</trace>\n    <trace>org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)</trace>\n    <trace>org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1624)</trace>\n    <trace>org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)</trace>\n    <trace>org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)</trace>\n    <trace>org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)</trace>\n    <trace>org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)</trace>\n    <trace>org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1594)</trace>\n    <trace>org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)</trace>\n    <trace>org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)</trace>\n    <trace>org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)</trace>\n    <trace>org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:59)</trace>\n    <trace>org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)</trace>\n    <trace>org.eclipse.jetty.server.Server.handle(Server.java:516)</trace>\n    <trace>org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:400)</trace>\n    <trace>org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:645)</trace>\n    <trace>org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:392)</trace>\n    <trace>org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)</trace>\n    <trace>org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)</trace>\n    <trace>org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)</trace>\n    <trace>org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:555)</trace>\n    <trace>org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:410)</trace>\n    <trace>org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:164)</trace>\n    <trace>org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)</trace>\n    <trace>org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)</trace>\n    <trace>org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)</trace>\n    <trace>org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)</trace>\n    <trace>org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)</trace>\n    <trace>org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)</trace>\n    <trace>org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)</trace>\n    <trace>org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)</trace>\n    <trace>org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)</trace>\n    <trace>java.lang.Thread.run(Thread.java:750)</trace>\n  </stackTrace>\n  <suppressedExceptions class=\"java.util.Collections$UnmodifiableRandomAccessList\" resolves-to=\"java.util.Collections$UnmodifiableList\">\n    <c class=\"list\"/>\n    <list/>\n  </suppressedExceptions>\n</com.mirth.connect.client.core.ControllerException>", :trace-redirects []}
	at slingshot.support$stack_trace.invoke(support.clj:201)
	at clj_http.client$exceptions_response.invokeStatic(client.clj:240)
	at clj_http.client$exceptions_response.invoke(client.clj:231)
	at clj_http.client$wrap_exceptions$fn__2402.invoke(client.clj:249)
	at clj_http.client$wrap_accept$fn__2612.invoke(client.clj:726)
	at clj_http.client$wrap_accept_encoding$fn__2619.invoke(client.clj:748)
	at clj_http.client$wrap_content_type$fn__2606.invoke(client.clj:709)
	at clj_http.client$wrap_form_params$fn__2705.invoke(client.clj:950)
	at clj_http.client$wrap_nested_params$fn__2726.invoke(client.clj:984)
	at clj_http.client$wrap_flatten_nested_params$fn__2735.invoke(client.clj:1008)
	at clj_http.client$wrap_method$fn__2673.invoke(client.clj:884)
	at clj_http.cookies$wrap_cookies$fn__1081.invoke(cookies.clj:131)
	at clj_http.links$wrap_links$fn__2207.invoke(links.clj:63)
	at clj_http.client$wrap_unknown_host$fn__2743.invoke(client.clj:1037)
	at clj_http.client$request_STAR_.invokeStatic(client.clj:1165)
	at clj_http.client$request_STAR_.invoke(client.clj:1158)
	at clj_http.client$post.invokeStatic(client.clj:1183)
	at clj_http.client$post.doInvoke(client.clj:1179)
	at clojure.lang.RestFn.invoke(RestFn.java:423)
	at mirthsync.http_client$post_xml.invokeStatic(http_client.clj:31)
	at mirthsync.http_client$post_xml.invoke(http_client.clj:21)
	at mirthsync.actions$upload_node.invokeStatic(actions.clj:19)
	at mirthsync.actions$upload_node.invoke(actions.clj:11)
	at mirthsync.actions$process_nodes.invokeStatic(actions.clj:95)
	at mirthsync.actions$process_nodes.invoke(actions.clj:83)
	at mirthsync.actions$upload.invokeStatic(actions.clj:118)
	at mirthsync.actions$upload.invoke(actions.clj:112)
	at mirthsync.apis$iterate_apis.invokeStatic(apis.clj:152)
	at mirthsync.apis$iterate_apis.invoke(apis.clj:146)
	at mirthsync.core$run$fn__3759.invoke(core.clj:25)
	at mirthsync.http_client$with_authentication.invokeStatic(http_client.clj:58)
	at mirthsync.http_client$with_authentication.invoke(http_client.clj:45)
	at mirthsync.core$run.invokeStatic(core.clj:21)
	at mirthsync.core$run.invoke(core.clj:10)
	at mirthsync.core$main_func$fn__3776.invoke(core.clj:46)
	at mirthsync.core$main_func.invokeStatic(core.clj:45)
	at mirthsync.core$main_func.doInvoke(core.clj:37)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at clojure.core$apply.invokeStatic(core.clj:665)
	at clojure.core$apply.invoke(core.clj:660)
	at mirthsync.core$_main.invokeStatic(core.clj:57)
	at mirthsync.core$_main.doInvoke(core.clj:55)
	at clojure.lang.RestFn.applyTo(RestFn.java:137)
	at mirthsync.core.main(Unknown Source)
clj-http: status 500`

**Comment by jessesaga** [comment]
One thing to be aware of is that mirthsync, currently, doesn't handle
renames or moving channels between groups very gracefully. I normally
delete my local target directory files (under version control) before doing
a pull after changing names or groups. Without deleting my local files I
would end up with 2 of the same channels in different locations after
changing names or moving groups. This would cause the behavior you're
seeing.

Additionally, to troubleshoot - you can look on the destination server via
Mirth Admin and look at the channel ID guids. You should be able to
identify the destination server channel and see which group it belongs to.
You could try to move the channel via mirth admin to the group matching the
group membership in the 'source' server'. That might fix the issue.

**Reply by anibal2j** [reply]
I noticed that my source mirth and my destination mirth were out of sync with respect to versions. I just did the upgrade on the destination mirth and ran the push again. Now I see two issues:

1. Not all groups were pushed over (some groups are there, but many are not)
2. Got the following errors/warnings when running mirthsynch push:

```
/usr/bin/java
found java executable in PATH
version 1.8.0_412
Authenticating to server at https://10.29.46.92:8443/api as service
	File: -f/ConfigurationMap.xml
Uploading from -f/ to /server/configurationMap
	File: -f/GlobalScripts/globalScripts.xml
Uploading from -f/GlobalScripts to /server/globalScripts
	File: -f/Resources.xml
Uploading from -f/ to /server/resources
	File: -f/CodeTemplates/AKPHS - PHS code only/index.xml
Uploading from -f/CodeTemplates to /codeTemplateLibraries
	File: -f/CodeTemplates/DigitalOne Outside Read Filters/index.xml
	File: -f/CodeTemplates/Utility functions/index.xml
WARN : An unexpected response was received from the server...
WARN : Status: 200, Phrase: OK, Body: <codeTemplateLibrarySaveResult>
  <overrideNeeded>true</overrideNeeded>
  <librariesSuccess>false</librariesSuccess>
  <libraryResults/>
  <codeTemplateResults/>
</codeTemplateLibrarySaveResult>
WARN : An unexpected response was received from the server...
WARN : Status: 200, Phrase: OK, Body: <codeTemplateLibrarySaveResult>
  <overrideNeeded>true</overrideNeeded>
  <librariesSuccess>false</librariesSuccess>
  <libraryResults/>
  <codeTemplateResults/>
</codeTemplateLibrarySaveResult>
	File: -f/CodeTemplates/AKPHS - PHS code only/save_mrn_acc_into_channel_map_akphs.xml
Uploading from -f/CodeTemplates to /codeTemplates
	File: -f/CodeTemplates/DigitalOne Outside Read Filters/is_inhouse_sunshine_order(send_email_on_error).xml
	File: -f/CodeTemplates/DigitalOne Outside Read Filters/is_inhouse_matsuhs_order(send_email_on_error).xml
	File: -f/CodeTemplates/DigitalOne Outside Read Filters/is_inhouse_medpark_order(send_email_on_error).xml
	File: -f/CodeTemplates/Utility functions/fixHL7NodeOrder(node).xml
	File: -f/CodeTemplates/Utility functions/date-related utilities.xml
	File: -f/CodeTemplates/Utility functions/save_mrn_acc_into_channel_map().xml
	File: -f/CodeTemplates/Utility functions/send_email(subject, body, email_addresses, org_code) %2F send_email_for_org(subject, body, org_code, map_of_emails, acc_number).xml
	File: -f/CodeTemplates/Utility functions/xFilter(xmlList, callback, minLimit, maxReturned, enforceMax).xml
	File: -f/CodeTemplates/Utility functions/getSerializedHl7(segment).xml
	File: -f/CodeTemplates/Utility functions/get_mapping(input_field_to_map, mapping_table, description).xml
	File: -f/CodeTemplates/Utility functions/postfix utilities.xml
	File: -f/Channels/VNA/index.xml
```

**Comment by jessesaga** [comment]
I believe that error is letting you know that the destination Mirth has a
more recent revision and it refuses to update unless you add the -f (force)
parameter.

Jesse Dowell

https://www.saga-it.com <https://saga-it.com/>

**Reply by anibal2j** [reply]
But I'm already using the -f option. so, why is it complaining?

**Comment by jessesaga** [comment]
Not sure. It should be sending the override parameter if you specify -f.
Are you using mirthsync 3.1.0?

**Reply by anibal2j** [reply]
yes, using 3.1.0. 

This is the pull command:

`/home/mirth/mirthsync/mirthsync/bin/mirthsync.sh -s https://localhost:8443/api -u service -p xxxxxxx -t "${git_dir}" -f pull -i --include-configuration-map
`
And this is the push command:

`/home/mirth/mirthsync/mirthsync/bin/mirthsync.sh -s https://slave:8443/api -u service -p xxxxxx -t "${git_dir}" -f push -i --include-configuration-map`

Should I be using something else instead? How can I help debug it?
